rustls/src/crypto/hpke.rs

```diff
@@ -0,0 +1,85 @@
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::fmt::Debug;
+
+use crate::msgs::enums::HpkeKem;
+use crate::msgs::handshake::HpkeSymmetricCipherSuite;
+use crate::Error;
+
+/// A provider for [RFC 9180] Hybrid Public Key Encryption (HPKE) in base mode.
+///
+/// At a minimum each provider must support the [HPKE ciphersuite profile] required for
+/// encrypted client hello (ECH):
+///  * KEM: DHKEM(X25519, HKDF-SHA256)
+///  * symmetric ciphersuite:  AES-128-GCM w/ HKDF-SHA256
+///
+/// [RFC 9180]: <https://www.rfc-editor.org/rfc/rfc9180.html>
+/// [HPKE ciphersuite profile]: <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#section-9>
+pub trait HpkeProvider: Debug + Send + Sync + 'static {
+    /// Create a base mode HPKE instance for the chosen suite.
+    ///
+    /// May return an error if the suite is unsupported by the provider.
+    fn create(&self, suite: &HpkeSuite) -> Result<Box<dyn Hpke>, Error>;
+
+    /// Does the provider support the given [HpkeSuite]?
+    fn supports_suite(&self, suite: &HpkeSuite) -> bool;
+}
+
+/// An HPKE suite, specifying a key encapsulation mechanism and a symmetric cipher suite.
+pub struct HpkeSuite {
+    /// The choice of HPKE key encapsulation mechanism.
+    pub kem: HpkeKem,
+
+    /// The choice of HPKE symmetric cipher suite.
+    ///
+    /// This combines a choice of authenticated encryption with additional data (AEAD) algorithm
+    /// and a key derivation function (KDF).
+    pub sym: HpkeSymmetricCipherSuite,
+}
+
+/// An HPKE instance that can be used for base-mode single-shot encryption and decryption.
+pub trait Hpke: Debug + Send + Sync {
+    /// Seal the provided `plaintext` to the recipient public key `pk_r` with application supplied
+    /// `info`, and additional data `aad`.
+    ///
+    /// Returns ciphertext that can be used with [Self::open] by the recipient to recover plaintext
+    /// using the same `info` and `aad` and the private key corresponding to `pk_r`.
+    fn seal(
+        &mut self,
+        pk_r: &HpkePublicKey,
+        info: &[u8],
+        aad: &[u8],
+        plaintext: &[u8],
+    ) -> Result<(EncapsulatedSecret, Vec<u8>), Error>;
+
+    /// Open the provided `ciphertext` using the encapsulated secret `enc`, with application
+    /// supplied `info`, and additional data `aad`.
+    ///
+    /// Returns plaintext if  the `info` and `aad` match those used with [Self::seal], and
+    /// decryption with `sk_r` succeeds.
+    fn open(
+        &mut self,
+        enc: &EncapsulatedSecret,
+        sk_r: &HpkePrivateKey,
+        info: &[u8],
+        aad: &[u8],
+        ciphertext: &[u8],
+    ) -> Result<Vec<u8>, Error>;
+}
+
+/// An HPKE public key.
+pub struct HpkePublicKey(pub Vec<u8>);
+
+/// An HPKE private key.
+pub struct HpkePrivateKey(pub Vec<u8>);
```

Comment:
The private key with public `Vec<u8>` field gives me a bit of pause. Should we wrap some API around this that requires being more intentional about how this gets accessed?

Reply:
> Should we wrap some API around this that requires being more intentional about how this gets accessed?

Do you mean something like a `secret_bytes(&self) -> &[u8]` accessor similar to how `pki_types::PrivatePkcs1KeyDer` has `secret_pkcs1_der`? Or did you have something more involved in mind?

Reply:
I've made some changes loosely modeled on `pki_types`. Hopefully it's closer to what you had in mind.

Reply:
Yup, I think this is good!